### PR TITLE
Fixing bug buy/sell limit orders

### PR DESCRIPTION
**Pull request by razine-bensari**
# Related Issue
- #312 

#### Proposed changes
- The issue was that we only placed market order and hardcoded the type. Now we support limit_buy, limit_sell, market_buy and market_sell

#### Additional Info
- for limit_* orders, the limit price needs to be provided in the paylod, the boolean for market hours. Please refer to the order object for field names

#### Checklist (if applicable)
- Did not break any unit tests. Please test locally if it works fine on your end. We stumbled upon a few issues where it was working well on some peoples machine but not all

#### Reviewer(s)
 - @DPigeon 
 - @Alessandro100 
 - @EspressoCode 
 - @abdulmansour 
 - @marwan-ayadi

### CODE REVIEW

**Comment by EspressoCode** [comment]
Right now we update the balance using the alpaca balance but they give us a negative balance, for example when you create a user on Alpaca, the cash balance is negative: "cash": "-1024.6538"

Our frontend has a $100,000 dollar balance for when the user has no purchases (which was hardcoded and not accurate!)
then we do the first purchase and it changes to (that negative value) - (price you bought), and then because you have a negative value the buy button becomes disabled.

I think we can not rely on Alpaca, especially that I saw that they update the balance sometimes on a daily basis! My solution is to set the balance equal to 100,000 for when we create the user, and then in each transaction, we manually deduct/add to the balance. What do you think? @Alessandro100 @razine-bensari @DPigeon @wltrf 
I can make those changes, but I mentioned it here so that we don't merge this pr until I add those changes here, because this bug is related to this issue.

**Comment by DPigeon** [comment]
> Right now we update the balance using the alpaca balance but they give us a negative balance, for example when you create a user on Alpaca, the cash balance is negative: "cash": "-1024.6538"
> 
> Our frontend has a $100,000 dollar balance for when the user has no purchases (which was hardcoded and not accurate!)
> then we do the first purchase and it changes to (that negative value) - (price you bought), and then because you have a negative value the buy button becomes disabled.
> 
> I think we can not rely on Alpaca, especially that I saw that they update the balance sometimes on a daily basis! My solution is to set the balance equal to 100,000 for when we create the user, and then in each transaction, we manually deduct/add to the balance. What do you think? @Alessandro100 @razine-bensari @DPigeon @wltrf
> I can make those changes, but I mentioned it here so that we don't merge this pr until I add those changes here, because this bug is related to this issue.

Good point. We need the user to have a 'base' balance. Totally agree.

File: backend/src/main/java/com/capstone/moneytree/service/impl/DefaultTransactionService.java [hunk 1]
```diff
@@ -126,6 +131,20 @@ private void executeTransaction(String alpacaKey, Order order, User user) {
       }
    }
 
+   private Order executeRequest(AlpacaAPI api, Order order) throws AlpacaAPIRequestException {
+      if (MoneyTreeOrderType.MARKET_BUY.name().equals(obtainMoneyTreeOrderType(order).name()) ||
```
Comment (DPigeon) [inline backend/src/main/java/com/capstone/moneytree/service/impl/DefaultTransactionService.java:135]: We'll have to test those two condition paths when the executeTransaction test is fixed.

Comment (razine-bensari) [reply backend/src/main/java/com/capstone/moneytree/service/impl/DefaultTransactionService.java]: Added extra tests cases

**Comment by razine-bensari** [comment]
> Right now we update the balance using the alpaca balance but they give us a negative balance, for example when you create a user on Alpaca, the cash balance is negative: "cash": "-1024.6538"
> 
> Our frontend has a $100,000 dollar balance for when the user has no purchases (which was hardcoded and not accurate!)
> then we do the first purchase and it changes to (that negative value) - (price you bought), and then because you have a negative value the buy button becomes disabled.
> 
> I think we can not rely on Alpaca, especially that I saw that they update the balance sometimes on a daily basis! My solution is to set the balance equal to 100,000 for when we create the user, and then in each transaction, we manually deduct/add to the balance. What do you think? @Alessandro100 @razine-bensari @DPigeon @wltrf
> I can make those changes, but I mentioned it here so that we don't merge this pr until I add those changes here, because this bug is related to this issue.

I agree, I actually dont like alpaca anymore, because of simple stuff like these. Its the little things man, the little things

**Comment by EspressoCode** [comment]
> > Right now we update the balance using the alpaca balance but they give us a negative balance, for example when you create a user on Alpaca, the cash balance is negative: "cash": "-1024.6538"
> > Our frontend has a $100,000 dollar balance for when the user has no purchases (which was hardcoded and not accurate!)
> > then we do the first purchase and it changes to (that negative value) - (price you bought), and then because you have a negative value the buy button becomes disabled.
> > I think we can not rely on Alpaca, especially that I saw that they update the balance sometimes on a daily basis! My solution is to set the balance equal to 100,000 for when we create the user, and then in each transaction, we manually deduct/add to the balance. What do you think? @Alessandro100 @razine-bensari @DPigeon @wltrf
> > I can make those changes, but I mentioned it here so that we don't merge this pr until I add those changes here, because this bug is related to this issue.
> 
> I agree, I actually dont like alpaca anymore, because of simple stuff like these. Its the little things man, the little things

After talking to Alessandro, I realized that the negative balance is because we are using money-tree users over the same Alpaca account in our local, so for prod users that wouldn't be an issue. I think we are good to merge this PR :-)

**Review by DPigeon — APPROVED** [review]
Approved. And yes Alpaca is a struggle now that we're experimenting it more.

**Pull request merged**